File: haskell/day2.hs
import Data.List


materialForPackages:: ((Int,Int,Int)->Int) -> String -> Int
materialForPackages material packages =
	sum $ map (material . parsePackage) $ lines packages

ribbon::(Int,Int,Int) -> Int
ribbon (x,y,z) = 2*x + 2*y + x*y*z


paper::(Int,Int,Int) -> Int
paper (x,y,z) = 2*x*y + 2*x*z + 2*y*z + x*y

parsePackage::String -> (Int, Int, Int)
parsePackage s = 
	let
		[x,y,z] = sort $ map read $ words $ map (\c -> if c == 'x' then ' ' else c) s
	in
		(x,y,z)


main = do packages <- readFile "../data/day2.txt";
		  putStrLn "****** Day 2 ******"
          putStrLn $ "The elves need " ++ (show $ materialForPackages paper packages) ++ " square feet of paper"
          putStrLn $ "They also need " ++ (show $ materialForPackages ribbon packages) ++ " feet of ribbon"

File: clojure/src/advent/day2.clj
(ns advent.day2
	(:require [clojure.string :refer [split]]))


(defn- parse-box [box]
	(->>
		box
		(re-seq #"\d+")
		(map read-string)
		sort))

(defn- paper-for-box [box]
	(let [[w l h] (parse-box box)]
		(+ (* 2 w l) (* 2 w h) (* 2 l h) (* w l))))

(defn- ribbon-for-box [box]
	(let [[w l h] (parse-box box)]
		(+ (* 2 w) (* 2 l) (* w l h) )))

(defn paper [boxes]
	(apply + (map paper-for-box (split boxes #"\n" ))))

(defn ribbon [boxes]
	(apply + (map ribbon-for-box (split boxes #"\n" ))))

(defn day2 []
	(let [boxes (slurp "../data/day2.txt")]
		(do
			(println "****** Day 2 *******")
			(println "The elves need" (paper boxes) "square feet of paper.")
			(println "They also need" (ribbon boxes) "feet of ribbon.")
			(println ))))




